### PR TITLE
MINIFICPP-1372 - Allow async content deletion

### PR DESCRIPTION
**Pull request by adamdebreceni**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/include/core/ContentRepository.h
Comment (lordgamez) [inline libminifi/include/core/ContentRepository.h:38]: Why is the inheritance from core::Repository needed for ContentRepository? There is a discussion about a common base class introduced in #1490 (https://github.com/apache/nifi-minifi-cpp/pull/1490#discussion_r1093458349). A suggested change is that core::Repository should be renamed as it represents the a key-value based repository while content repo has a stream based interface so it should not be a common base class. This will probably conflict with #1485 maybe we should try rebasing this change on that.

Comment (adamdebreceni) [reply libminifi/include/core/ContentRepository.h:38]: I am fine with waiting until the mentioned PRs are merged

Comment (adamdebreceni) [reply libminifi/include/core/ContentRepository.h]: in light of changes since and future plans, rewrote this PR to not use inheritance

File: extensions/rocksdb-repos/DatabaseContentRepository.cpp
Comment (szaszm) [inline extensions/rocksdb-repos/DatabaseContentRepository.cpp:259]: We should handle the case when there is an interruption (kill, shutdown, etc) between adding the keys to the vector, and actually deleting them, including an interruption between these lines: emptying `keys_to_delete_` and actually deleting. For the latter part, doing the delete before emptying the list, but while holding the mutex may do the trick, but may not be the best idea from a performance standpoint. What do you think?

Do we store `keys_to_delete_` on disk?

**Review by lordgamez — APPROVED** [review]

File: extensions/rocksdb-repos/DatabaseContentRepository.cpp
Comment (fgerlits) [inline extensions/rocksdb-repos/DatabaseContentRepository.cpp:234]: Nitpicking, but I would make the inside of this `if` into a separate function.  This way, we have to make sure to have a `return` at the end of every code path; that way, the compiler would check it for us.

**Review by fgerlits — COMMENTED** [review]
LGTM, just some minor comments

File: conf/minifi.properties
Comment (adamdebreceni) [reply conf/minifi.properties:41]: based on the time parsing logic, it is not allowed, should we explicitly check for "0" or rather some other value? (e.g. "sync")

File: extensions/rocksdb-repos/DatabaseContentRepository.cpp
Comment (adamdebreceni) [reply extensions/rocksdb-repos/DatabaseContentRepository.cpp:259]: on a restart we first comb through the content repository and delete all content without a flowfile referencing them (in `clearOrphans`) that should prevent leaking content

File: extensions/rocksdb-repos/DatabaseContentRepository.cpp
Comment (adamdebreceni) [reply extensions/rocksdb-repos/DatabaseContentRepository.cpp]: done

File: CONFIGURE.md
Comment (adamdebreceni) [reply CONFIGURE.md]: done

File: libminifi/test/rocksdb-tests/DBContentRepositoryTests.cpp
Comment (adamdebreceni) [reply libminifi/test/rocksdb-tests/DBContentRepositoryTests.cpp:131]: it wasn't safe as the gc runs first before waiting, flipped it so that the gc does not run at 0, so it should be safe now

File: conf/minifi.properties
Comment (fgerlits) [reply conf/minifi.properties:41]: I'm fine with any of these options, as long as the documentation is in line with the reality.  Maybe we could even update the time parsing logic to allow "0"?

Comment (adamdebreceni) [reply conf/minifi.properties]: added explicit check for "0" while we sort out if we want to allow parsing "0" as time in all places (in probably a separate PR)

File: extensions/rocksdb-repos/DatabaseContentRepository.cpp
Comment (adamdebreceni) [reply extensions/rocksdb-repos/DatabaseContentRepository.cpp:127]: as I understand, compaction and deletion could run concurrently without blocking each other, they also operate on different time scales, merging them could be troublesome

**Review by fgerlits — APPROVED** [review]

**Review by martinzink — APPROVED** [review]

File: extensions/rocksdb-repos/DatabaseContentRepository.cpp
Comment (adamdebreceni) [reply extensions/rocksdb-repos/DatabaseContentRepository.cpp:268]: added

**Review by szaszm — APPROVED** [review]

**Pull request closed without merging**
